### ISSUE TITLE
Epic: on-demand S3 downloads

### ISSUE THREAD

**Issue by kelvich** [neondatabase/neon#2029]
# On-Demand Download

This epic tracks the effort to get Heikki's https://github.com/neondatabase/neon/pull/2595 shipped.

## Roadmap

(Dates in this list is time of merge into `main`)

1. Nov 22 -> Nov 25 -> Nov 28: Land `storage_sync2` PR #2785
    - [x] root out flaky tests
    - [x] port relevant unresolved conversations in original PR https://github.com/neondatabase/neon/pull/2785#issuecomment-1324041213
    - [x] scan for `TODO` / `FIXME` and convert into follow-up issues
      - on Nov 23, @problame opened corresponding review comments for all FIXMEs. Will need another pass once those are addressed.
    - [x] update `storage_sync` doc comment https://github.com/neondatabase/neon/pull/2903
    - [x] test coverage for retrying uploads https://github.com/neondatabase/neon/pull/2892
        - not needed for downloads at this point because
            - Tenant::attach() can be retried manually, and it will resume attach correctly
            - once on-demand download lands, we'll have retries in there
    - [x] address FIXMEs regarding race conditions around setting tenant state, e..g, concurrent attach & detach #2910
    - [x] handle & regress-test race condition between timeline delete & layer file upload https://github.com/neondatabase/neon/pull/2921
1. After `storage_sync2` lands in staging: exercise `storage_sync2` attach code — @LizardWizzard, possibly, can be affected by these changes with the migration. Nevertheless, he will touch these parts during the migration.
    - kill a PS and create a new one
    - manual detach & attach of individual tenants
    - together with Kirill / SRE / on-call people
    - make sure existing tools work (`neonadmin`)
1. Nov 25 -> Dec 1 : Follow-up changes to `storage_sync2`
    - [x] #2977 **Blocker for production**
    - [x] #3006 **~~Blocker for production~~**

    - [x] bring back metrics (`measure_remote_op`) https://github.com/neondatabase/neon/pull/2892
    - [x] #2928
    - [x] #2946
    - [x] #2930 
    - [x] #3018
    - [x] #2907
    - [x] **Make sure tenant reload can be used to resolve clogged upload queue #2537** Duplicates #2977 
1. Dec 16 -> Dec 21: Port On-Demand Download https://github.com/neondatabase/neon/pull/3013
 
1. Dec 20 -> Dec 31: Refactor code to reflect new separation of concerns & reduce source-level conflicts in the future.
 - #3210
 - ~~Move tenant_* files inside this module~~.
 - ~~Split storage_sync.rs into upload_queue and remote_timeline_client~~
    

6. Jan 6: Stop/delete/start the project on stage to test the on-demand download 
- [x] #3249

7. #3287
---

Original description from @kelvich 

Right now we download all the history from the S3 during tenant attach. Way better to do that on-demand.

As we will have on-demand load we can think of cold data eviction (probably as a separate issue).

- [x] https://github.com/neondatabase/neon/issues/2035
- [x] https://github.com/neondatabase/neon/pull/2376
- [x] https://github.com/neondatabase/neon/pull/2231
- [x] https://github.com/neondatabase/neon/issues/2456

**Comment by hlinnaka** [comment]
>  port relevant unresolved conversations in original PR

I looked at the the unresolved conversations in https://github.com/neondatabase/neon/pull/2595 and https://github.com/neondatabase/neon/pull/2595, and didn't see anything that needs action right now. Ticked that box

**Comment by shanyp** [comment]
> Follow-up epic: #3208 this one can be closed?

Add your Issue to be part of this Epic
